src/storage/mod.rs

```diff
@@ -129,6 +140,145 @@ impl Storage {
+        // since we only have a range anyways, and we need the encoding
+        // for the range, not the file
+        if let Some(alg) = compression {
+            blob.content = decompress(blob.content.as_slice(), alg, max_size)?;
+            blob.compression = None;
+        }
+        Ok(blob)
+    }
+
+    fn get_index_for(&self, archive_path: &str) -> Result<archive_index::Index, Error> {
+        let remote_index_path = format!("{}.index", archive_path);
+        let local_index_path = self.local_archive_cache_path.join(&remote_index_path);
+
+        if local_index_path.exists() {
+            let mut file = fs::File::open(local_index_path)?;
+            archive_index::Index::load(&mut file)
+        } else {
+            let index_content = self.get(&remote_index_path, std::usize::MAX)?.content;
+
+            fs::create_dir_all(
+                local_index_path
+                    .parent()
+                    .ok_or_else(|| err_msg("index path without parent"))?,
+            )?;
+            let mut file = fs::File::create(&local_index_path)?;
+            file.write_all(&index_content)?;
+
+            archive_index::Index::load(&mut &index_content[..])
+        }
+    }
+
+    pub(crate) fn get_from_archive(
+        &self,
+        archive_path: &str,
+        path: &str,
+        max_size: usize,
+    ) -> Result<Blob, Error> {
+        let index = self.get_index_for(archive_path)?;
+        let info = index.find_file(path)?;
+
+        let blob = self.get_range(
+            archive_path,
+            max_size,
+            info.range(),
+            Some(info.compression()),
+        )?;
+
+        Ok(Blob {
+            // TODO: what is this path used for after `get`? Can we really just concat zip and
+            // path inside?
+            path: format!("{}/{}", archive_path, path),
+            mime: detect_mime(&path).into(),
+            date_updated: blob.date_updated,
+            content: blob.content,
+            compression: None,
+        })
+    }
+
+    pub(crate) fn store_all_in_archive(
+        &self,
+        archive_path: &str,
+        root_dir: &Path,
+    ) -> Result<(HashMap<PathBuf, String>, CompressionAlgorithm), Error> {
+        let mut file_paths = HashMap::new();
+
+        // due to how we retrieve the archived files, the compression method has to be supported
+        // by our normal content-encoding / compression logic.
+        // Deflate is not supported right now, and is compressing worse anyways.
+        let options =
+            zip::write::FileOptions::default().compression_method(zip::CompressionMethod::Bzip2);
+
+        let mut zip = zip::ZipWriter::new(io::Cursor::new(Vec::new()));
+        for file_path in get_file_list(root_dir)? {
+            let mut file = fs::File::open(root_dir.join(&file_path))?;
+
+            // TODO: should we create directories with `start_directory`?
+            // it's not standard, but convention for zip files
```

Comment:
There is a little more info [here](https://github.com/thejoshwolfe/yauzl/issues/52#issuecomment-287568422). 

In short: The ZIP format only supports file-entries, and not directories. The files can have a path, and will be extracted to these folders, so this is only a problem with empty directories in the archive. 

There is a convention to use (kind of) empty file entries with a `/` at the end, and most (if not all) libraries follow this. 

Currently I only create the file-entries, and I don't think we need the separate directory entries, as long we don't need empty folders. 

We _could_, for the take of following the convention, also create the directory entries, personally I don't think we need to.

Reply:
Oh I see - I don't think we even *can* serve empty directories, it will always try to look for index.html and give an error otherwise. I agree I don't think we should store them in the zipfile.